lib/ts/controllers/s-modal.ts

```diff
@@ -0,0 +1,222 @@
+            this._boundTabTrap = this._boundTabTrap || ((e: KeyboardEvent) => {
+                // if somehow the user has tabbed out of the modal or if focus started outside the modal, push them to the first item
+                if (!this.modalTarget.contains(<Element>e.target)) {
+                    e.preventDefault();
+                    firstTabbable.focus();
+                }
+
+                // if they've tabbed backwards over the first item, then go to the last item
+                if (e.target == firstTabbable && e.keyCode === 9 && e.shiftKey) {
+                    e.preventDefault();
+                    lastTabbable.focus();
+                }
+
+                // if they've tabbed forwards over the last item, then go to the first item
+                if (e.target == lastTabbable && e.keyCode === 9 && !e.shiftKey) {
+                    e.preventDefault();
+                    firstTabbable.focus();
+                }
+            });
+
+            document.addEventListener("keydown", this._boundTabTrap);
+        }
+
+        /**
+         * Binds global events to the document for hiding popovers on user interaction
+         */
+        _bindDocumentEvents () {
+            // in order for removeEventListener to remove the right event, this bound function needs a constant reference
+            this._boundClickFn = this._boundClickFn || this._hideOnOutsideClick.bind(this);
+            this._boundKeypressFn = this._boundKeypressFn || this._hideOnEscapePress.bind(this);
+
+            document.addEventListener("click", this._boundClickFn);
+            document.addEventListener("keyup", this._boundKeypressFn);
+
+            this._bindTabFocusTrap();
+        }
+
+        /**
+         * Unbinds global events to the document for hiding popovers on user interaction
+         */
+        _unbindDocumentEvents () {
+            document.removeEventListener("click", this._boundClickFn);
+            document.removeEventListener("keyup", this._boundKeypressFn);
+            document.removeEventListener("keydown", this._boundTabTrap);
+        }
+
+        /**
+         * Forces the popover to hide if a user clicks outside of it or its reference element
+         */
+        _hideOnOutsideClick (e: Event) {
+            var target = <Node>e.target;
+            // check if the document was clicked inside either the toggle element or the modal itself
+            // note: .contains also returns true if the node itself matches the target element
+            if (!this.modalTarget.querySelector(".s-modal--dialog")!.contains(target)) {
+                this._toggle(false);
+            }
+        }
+
+        /**
+         * Forces the popover to hide if the user presses escape while it, one of its childen, or the reference element are focused
+         */
+        _hideOnEscapePress (e: KeyboardEvent) {
```

Comment:
If we have a popover inside a modal, `Esc` will kill the outer rather than the inner.  This needs to be more sophisticated if we're binding at the document level.

Reply:
Very good point. I'm not sure off-hand how I'd do it other than manually checking if the modal has esc closeable items... I'm going to punt on this now so we can get this in, but I'll be thinking about it over the next few days.